### ISSUE TITLE
Memory Leak in BinarPrecisionRecallCurve

### ISSUE THREAD

**Issue by suahelen** [Lightning-AI/torchmetrics#2921]
Hi Folks,

I stumbled over this issue while playing around with anomlib
## 🐛 Bug
See this issue in anomalib:
https://github.com/openvinotoolkit/anomalib/issues/2531

To keep it short continuously calling `compute` on `BinaryPrecisionRecallCurve(Metric)` grows the `self.preds` and `self.target` indefinitely. 

This eventually leads to a memory overflow.

the following code  resolves the issue by clearing the preds and targets after concatenating them to the state
```py
    def compute(self) -> tuple[Tensor, Tensor, Tensor]:
        """Compute metric."""
        if self.thresholds is None:
            if not self.preds or not self.target:
                return torch.tensor([]), torch.tensor([]), torch.tensor([])
            state = (torch.cat(self.preds), torch.cat(self.target))

            self.preds.clear()
            self.target.clear()
        else:
            state = self.confmat
            self.confmat.zero_()

        precision, recall, thresholds = _binary_precision_recall_curve_compute(state, self.thresholds)
        return precision, recall, thresholds if thresholds is not None else torch.tensor([])
```

Cheers

**Comment by rittik9** [comment]
@Borda I would be interested in working on this

**Comment by SkafteNicki** [comment]
Hi all, thanks for reporting this issue.
I have a hard time understanding what the actual problem seems to be because what is reported here seems to be standard torchmetrics behavior. Metric states are not reset unless this is explicitly asked for by the user by calling the `.reset()` method. Calling `.compute()` does not reset the states, never have. Why is this, well there are usecase where users call update, ..., update -> compute -> update, ..., update -> compute and want to keep accumulate while still getting intermediate results. Therefore, if one want to clear the states after `.compute()` just call `.reset()`  afterwards.

Now the actual claim in this issue is that repeated call to `.compute()` increases memory, which I cannot reproduce. Here are some test code:
```python
from torchmetrics.classification import BinaryPrecisionRecallCurve
import torch
import gc

def allocated_tensors():
    for obj in gc.get_objects():
        try:
            if torch.is_tensor(obj) or (hasattr(obj, 'data') and torch.is_tensor(obj.data)):
                print(type(obj), obj.size())
        except:
            pass

def mem():
    return torch.cuda.memory_allocated()

preds = torch.rand(100).to(torch.device("cuda"))
target = torch.randint(0, 2, (100,)).to(torch.device("cuda"))
print(f"Memory before metric init: {mem():.2f} MB")


precision_recall_curve = BinaryPrecisionRecallCurve(compute_with_cache=False).to(torch.device("cuda"))
precision_recall_curve.update(preds, target)

print(f"Memory after update: {mem():.2f} MB")
allocated_tensors()

for i in range(5):
    precision_recall_curve.compute()
    print(f"Memory after compute {i}: {mem():.2f} MB")
    allocated_tensors()
```
Running this you will see something like this:
![Image](https://github.com/user-attachments/assets/5250abfe-4f80-438e-b6dc-1695d070136c)
So the output should be read like this:
* tensors are initialized and memory increases
* update is called and memory increases a bit due to bookkeeping inside the metric, but mainly it is still the tensors
* 3 tensors are printed, which is the `preds`  and `target` + a extra copy of `preds` due to a [torch.where](https://github.com/Lightning-AI/torchmetrics/blob/15dd9831278d5424fbab7967fdcf56134eb27618/src/torchmetrics/utilities/compute.py#L225-L229) which apparently prevents proper cleanup. That is a separate issue that I am first aware of now.
* After the first compute, memory increases and three new tensors are added. This is the resulting `precision`, `recall` and `threshold` tensor of the `.compute()` method. Why are these present if the results are not saved to some output variable? it is because the result of `.compute` is cached to make sure repeated calls to `.compute` without calling `.update` in-between is "free". Since data has not changed, the output should not change. To disable the cache, one could set `BinaryPrecisionRecallCurve(compute_with_cache=False)` on metric init (global arg for all metrics) and re-running the script would result in the output tensors no longer being in memory.
* The following repeated calls to `compute` does not change memory or allocated tensors, because nothing new is stored and in fact it is just a cache that is retrieved.

So I cannot really see that repeated calls to compute increases memory.

**Comment by rittik9** [comment]
@SkafteNicki yes you are right. Repeated call to `.compute()` does not increase memory. Calling `.update()` repeatedly increases memory. which can be fixed just by calling `.reset()` afterwards. 

![Image](https://github.com/user-attachments/assets/da8e54d4-4b75-45ff-a9f7-132de614267f)

![Image](https://github.com/user-attachments/assets/97f1ea1f-7d87-472a-951a-140f80fc5864)

**Comment by SkafteNicki** [comment]
@suahelen thanks for providing some more info. I can now confirm that the problem is on anomlib and not torchmetrics side. The [OneClassPostProcessor](https://github.com/openvinotoolkit/anomalib/blob/main/src/anomalib/post_processing/one_class.py) you link to log the metrics like this:

https://github.com/openvinotoolkit/anomalib/blob/6a6a80c1279e911c107607523b0655c67c763213/src/anomalib/post_processing/one_class.py#L110-L125

they are only calling `.compute` on the metrics and not `.reset` afterwards. The relevant documentation can be found here, https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html#logging-torchmetrics, what they are doing is what we are calling *manual* logging (=reset not handled by lightning) and `.reset()` should therefore manually be called. 

For completness, the linked code should be changed to:
```python
    def on_validation_epoch_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
        """Compute final threshold and normalization values.

        Args:
            trainer (Trainer): PyTorch Lightning trainer instance.
            pl_module (LightningModule): PyTorch Lightning module instance.
        """
        del trainer, pl_module
        if self._image_threshold_metric.update_called:
            self._image_threshold = self._image_threshold_metric.compute()
            self._image_threshold_metric.reset()
        if self._pixel_threshold_metric.update_called:
            self._pixel_threshold = self._pixel_threshold_metric.compute()
            self._pixel_threshold_metric.reset()
        if self._image_min_max_metric.update_called:
            self.image_min, self.image_max = self._image_min_max_metric.compute()
            self._image_min_max_metric.reset()
        if self._pixel_min_max_metric.update_called:
            self.pixel_min, self.pixel_max = self._pixel_min_max_metric.compute()
            self._pixel_min_max_metric.reset()
```
and similar for other `_on_epoch_end` methods.

Closing the issue on our side.

**Comment by suahelen** [comment]
Perfect thanks for the feedback,
I'll forward this information.
